### PR TITLE
ft(S3C-3006): Add metric calculation macro

### PR DESCRIPTION
**Pull request by tmacro**
Add the metric calculation macros
Small logic fixes in `createCheckpoint` and `createSnapshot`
Some general cleanup

 Note on `LOGMSG`:
`LOGMSG` requires a specific debugging extension (the one commented out in the Dockerfile) which is not meant to be used in production. I've left them commented to allow a developer to enable them as needed.

### CODE REVIEW

File: libV2/constants.js [hunk 1]
```diff
@@ -20,17 +20,21 @@ module.exports = {
         'deleteBucket',
         'deleteBucketCors',
         'deleteBucketReplication',
+        'deleteBucketReplication',
```
Comment (anurag4DSB) [inline libV2/constants.js]: `deleteBucketReplication` is twice, may be a left over?

File: tests/functional/v2/server/testListMetrics.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    body[level] = resources;
+
+    return needle(
+        'post',
+        `http://localhost:8100/${level}`,
+        body,
+        { json: true },
+    );
+}
+
```
Comment (anurag4DSB) [inline tests/functional/v2/server/testListMetrics.js]: extra lines(not sure if your linter catches it)

File: tests/functional/v2/server/testListMetrics.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+function opsToResp(operations) {
+    return Object.entries(operations)
+        .reduce((prev, [key, value]) => {
+            prev[operationToResponse[key]] = value;
+            return prev;
+        }, { ...emptyOperationsResponse });
+}
+
+
```
Comment (anurag4DSB) [inline tests/functional/v2/server/testListMetrics.js]: extra lines(not sure if your linter catches it)

File: tests/utils/v2Data/events.js [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+        'abortMultipartUpload',
+        'completeMultipartUpload',
+        'initiateMultipartUpload',
+        'listMultipartUploadParts',
+        'listMultipartUploads',
+        'createBucket',
+        'deleteBucket',
+        'deleteBucketCors',
+        'deleteBucketReplication',
+        'deleteBucketReplication',
```
Comment (anurag4DSB) [inline tests/utils/v2Data/events.js]: same here `deleteBucketReplication` is twice. I am now thinking if its at 2 places, is there a reason behind it?

**Review by anurag4DSB — CHANGES_REQUESTED** [review]
I could find some minor possible issues. Have no clue about the *.mc2 files yet :)

File: tests/functional/v2/server/testListMetrics.js (hunk 2 above)
Comment (tmacro) [reply tests/functional/v2/server/testListMetrics.js]: Technically 2 empty lines aren't an error, but consistency is better :)

File: tests/utils/v2Data/events.js (hunk 4 above)
Comment (tmacro) [reply tests/utils/v2Data/events.js]: copy/paste

File: tests/functional/v2/server/testListMetrics.js (hunk 2 above)
Comment (anurag4DSB) [reply tests/functional/v2/server/testListMetrics.js]: You code is too clean, I couldn't resist. :D
I majorly wanted to raise the 2 deleteBucketReplication thing :)

File: tests/utils/v2Data/events.js (hunk 4 above)
Comment (anurag4DSB) [reply tests/utils/v2Data/events.js]: understood, copy pasting variables.

**Comment by tmacro** [comment]
As heads up, due to some fat fingered-ness #558 was mistakenly merged into this PR. Come chaos may ensue as I fix it

**Comment by miniscruff** [comment]
> As heads up, due to some fat fingered-ness #558 was mistakenly merged into this PR. Come chaos may ensue as I fix it

:boom:

**Review by anurag4DSB — APPROVED** [review]

File: warpscript/utapi/getMetricsAt.mc2 [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+    '>
+  ]
+} 'info' STORE
+
+<%
+  !$info INFO
+  SAVE 'context' STORE
+  <%
+    'getMetrics' SECTION
```
Comment (jonathan-gramain) [inline warpscript/utapi/getMetricsAt.mc2]: is `getMetrics` what you want, or did you intend to put `getMetricsAt` here?

File: warpscript/util/getDefault.mc2 [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    RETHROW
+  %>
+  <% // finally, restore the context
+    $context RESTORE
+  %> TRY
+%>
+'macro' STORE
+
+// Unit tests
+// { 'a' 1 } 'a' 0 @macro 1 == ASSERT
```
Comment (jonathan-gramain) [inline warpscript/util/getDefault.mc2]: Shouldn't the unit tests be uncommented?

**Review by nicolas2bert — APPROVED** [review]

**Pull request merged**